aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1499,48 +1496,60 @@ std::tuple<Tensor, Tensor, Tensor> _lu_with_info_cpu(const Tensor& self, bool pi
-  auto batch_size = batchCount(A);
-  auto n = A.size(-2);
-  auto nrhs = b.size(-1);
-  auto lda = std::max<int64_t>(1, n);
-  auto infos_data = infos.data_ptr<int>();
+  TORCH_INTERNAL_ASSERT(input.device() == other.device());
+  TORCH_INTERNAL_ASSERT(input.device() == result.device());
+  TORCH_INTERNAL_ASSERT(input.device() == clone_input.device());
+  TORCH_INTERNAL_ASSERT(input.device() == infos.device());
 
-  for (const auto i : c10::irange(batch_size)) {
-    scalar_t* A_working_ptr = &A_data[i * A_mat_stride];
-    scalar_t* b_working_ptr = &b_data[i * b_mat_stride];
-    int* infos_working_ptr = &infos_data[i];
-    lapackTriangularSolve<scalar_t>(uplo, trans, diag, n, nrhs, A_working_ptr, lda, b_working_ptr, lda, infos_working_ptr);
+  TORCH_INTERNAL_ASSERT(input.scalar_type() == other.scalar_type());
+  TORCH_INTERNAL_ASSERT(input.scalar_type() == result.scalar_type());
+  TORCH_INTERNAL_ASSERT(input.scalar_type() == clone_input.scalar_type());
+
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
+  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
+
+  // if 'result' has no elements we can modify it
+  if (result.numel() == 0) {
+    result.resize_(other.transpose(-2, -1).sizes(), MemoryFormat::Contiguous);
+    result.transpose_(-2, -1);  // make 'result' to have Fortran contiguous memory layout
   }
-#endif
-}
 
-std::tuple<Tensor, Tensor> _triangular_solve_helper_cpu(const Tensor& self, const Tensor& A,
-                                                        bool upper, bool transpose, bool unitriangular) {
-  auto self_working_copy = cloneBatchedColumnMajor(self);
-  auto A_working_copy = cloneBatchedColumnMajor(A);
-  auto infos_tensor = at::empty(batchCount(self), self.options().dtype(kInt));
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(self.scalar_type(), "triangular_solve_cpu", [&]{
-    apply_triangular_solve<scalar_t>(self_working_copy, A_working_copy, infos_tensor, upper, transpose, unitriangular);
-  });
-  if (self.dim() > 2) {
-    batchCheckErrors(infos_tensor, "triangular_solve_cpu");
-  } else {
-    singleCheckErrors(infos_tensor.item<int64_t>(), "triangular_solve_cpu");
+  // if 'clone_input' has no elements we can modify it
+  if (clone_input.numel() == 0) {
+    clone_input.resize_(input.transpose(-2, -1).sizes(), MemoryFormat::Contiguous);
+    clone_input.transpose_(-2, -1);  // make 'clone_input' to have Fortran contiguous memory layout
   }
-  return std::tuple<Tensor, Tensor>(self_working_copy, A_working_copy);
+
+  // 'result' and 'clone_input' must be in batched column major order (Fortran contiguous)
+  TORCH_INTERNAL_ASSERT(result.transpose(-2, -1).is_contiguous());
+  TORCH_INTERNAL_ASSERT(clone_input.transpose(-2, -1).is_contiguous());
+
+  // triangular_solve_stub performs calculations in-place
+  // 'result' must be a copy of 'other'
+  // 'clone_input' must be a copy of 'input'
+  result.copy_(other);
```

Comment:
What happens if result is non-empty but is the incorrect size? Same question with clone_input.

Reply:
This can't happen with the current code, because the out= variant is implemented via a temporary tensor for the result and copy. Added a fusing copy path can be made in other PR. I didn't include it here because it'd too many changes for the current PR.

I'll add the assert for the sizes.